feat: 일정 CRUD 기능 구현

jooyeon0907 opened the pull request:
### 변경사항
<!-- 이 PR에서 어떤점들이 변경되었는지 기술해주세요. 가급적이면 as-is, to-be를 활용해서 작성해주세요.  -->
**AS-IS**
- 일정 관련 기능이 구현되지 않음

**TO-BE**
- 일정 CRUD 기능 구현
    - 일정 생성, 조회, 수정, 삭제 기능을 추가
- 기간별 일정 목록 조회
    - date 는 기준 날짜, period 는 monthly, weekly, daily 로 요청 가능 
    - 월간 조회
        - 예시: date = 2025-03-08, period = monthly
        - 조회 범위: 2025-03-01 ~ 2025-03-31 (해당 월의 전체 일정)
    - 주간 조회
        - 예시: date = 2025-03-08 (토요일), period = weekly
        - 조회 범위: 2025-03-02 (일) ~ 2025-03-08 (토) 
    - 일일 조회
        - 예시: date = 2025-03-08, period = daily
        - 조회 범위: 2025-03-08 하루 일정을 시간 별 조회
- GlobalExceptionHandler에 ScheduleException 추가
    - 일정 관련 예외 처리를 전역적으로 관리

### 테스트
<!-- 본 변경사항이 테스트가 되었는지 기술해주세요 --> 
- [ ] 테스트 코드
- [x ] API 테스트  (Postman)
    - 일정 추가 요청
        - 유효성 검사 확인
            - 필수 값: title, startDate, endDate
            - endDate 를 startDate 이전으로 설정할 수 없음
        - 예외 처리
            - 인증되지 않는 JWT 토큰: "로그인 정보가 유효하지 않습니다."
            - 캘린더 멤버가 아님: "해당 사용자는 이 캘린더의 멤버가 아닙니다."
        - 일정 생성 성공 여부 확인
    - 일정 목록 조회 요청
        - 예외 처리
            - 인증되지 않는 JWT 토큰: "로그인 정보가 유효하지 않습니다."
            - 캘린더 멤버가 아님: "해당 사용자는 이 캘린더의 멤버가 아닙니다."
            - 인증되지 않은 period 값:"period는 monthly, weekly, 또는 daily 중 하나여야 합니다."
        - 일정 목록 조회 확인
    - 일정 조회 요청
        - 예외 처리
            - 인증되지 않는 JWT 토큰: "로그인 정보가 유효하지 않습니다."
            - 캘린더 멤버가 아님: "해당 사용자는 이 캘린더의 멤버가 아닙니다."
            - 존재하지 않는 일정: "일정 정보가 존재하지 않습니다."
        - 일정 조회 성공 여부 확인
    - 일정 수정 요청
        - 유효성 검사 확인
            - endDate 를 startDate 이전으로 설정할 수 없음
        - 예외 처리
            - 인증되지 않는 JWT 토큰: "로그인 정보가 유효하지 않습니다."
            - 캘린더 멤버가 아님: "해당 사용자는 이 캘린더의 멤버가 아닙니다."
            - 존재하지 않는 일정: "일정 정보가 존재하지 않습니다."
        - 일정 수정 성공 여부 확인
    - 일정 삭제 요청
        - 예외 처리
            - 인증되지 않는 JWT 토큰: "로그인 정보가 유효하지 않습니다."
            - 캘린더 멤버가 아님: "해당 사용자는 이 캘린더의 멤버가 아닙니다."
        - 일정 삭제 성공 여부 확인

jooyeon0907 committed:
Merge remote-tracking branch 'origin/feature/calendar_member' into feature/schedule

jooyeon0907 committed:
feat: 일정 CRUD 기능 구현

jooyeon0907 committed:
feat: 일정 CRUD 기능 구현

jooyeon0907 committed:
Merge branch 'develop' of https://github.com/jooyeon0907/timate into feature/schedule

jooyeon0907 committed:
fix: 코드 수정

jooyeon0907 committed:
fix: 코드 수정

mingming-mentor commented on src/main/java/com/zerobase/timate/dto/ScheduleDto.java:
```diff
@@ -0,0 +1,110 @@
+		private String address;
+		@JsonProperty("latitude")
+		private double latitude;
+		@JsonProperty("longitude")
```
@JsonProperty의 값과 필드명이 같으면 굳이 넣어주지 않아도 동작은 제대로 될것 같아요

mingming-mentor commented on src/main/java/com/zerobase/timate/dto/ScheduleDto.java:
```diff
@@ -0,0 +1,110 @@
+	@NoArgsConstructor
+	@AllArgsConstructor
+	@Builder
+	public static class Response {
```
snake case 대응으로 @JsonProperty를 사용하신 거라면 @JsonNaming(value = PropertyNamingStrategies.SnakeCaseStrategy.class) 을 사용하셔도 좋을 것 같네요

jooyeon0907 replied:
초반에 직렬화 문제로 사용했던거 같은데 잘 기억이 안 나네요..ㅜㅜ 지금은 @JsonProperty 를 빼도 정상 동작되는 것을 확인하여 @JsonProperty 사용된 곳들 다 제거하겠습니다!

mingming-mentor commented on src/main/java/com/zerobase/timate/service/ScheduleService.java:
```diff
@@ -0,0 +1,128 @@
+		LocalDate startDate;
+		LocalDate endDate;
+
+		switch (period.toLowerCase()) {
```
period를 enum으로 관리해도 좋을 것 같아요. String이면 형변환이 계속 일어나야하고 string끼리의 비교가 계속 일어나므로 enum이 좀더 효율적이고 깔끔하게 코드 관리가 가능할 것 같아요

mingming-mentor commented on src/main/java/com/zerobase/timate/service/ScheduleService.java:
```diff
@@ -0,0 +1,128 @@
+
+		commonService.checkCalendarMember(userId, calendarId);
+
+		LocalDate startDate;
+		LocalDate endDate;
+
+		switch (period.toLowerCase()) {
+			case "monthly": // 월간 조회
+				startDate = referenceDate.withDayOfMonth(1);
+				endDate = referenceDate.withDayOfMonth(referenceDate.lengthOfMonth());
+				break;
+			case "weekly": // 주간 조회 (일요일 ~ 토요일 기준)
+				startDate = referenceDate.with(TemporalAdjusters.previousOrSame(DayOfWeek.SUNDAY));
+				endDate = startDate.plusDays(6);
+				break;
+			case "daily": // 일일 조회
+				startDate = referenceDate;
+				endDate = referenceDate;
+				break;
+			default:
+				throw new ScheduleException(INVALID_PERIOD);
+		}
```
이부분은 startDate, endDate를 구하는 부분으로 보이는데요, switch 문으로 인해서 startDate, endDate가 선언과 할당이 동시에 이루어지지 않는 것같아요. 이 부분은 안티패턴이므로 피해야합니다.
방법은 두가지가 있는데요, 우선 첫번째로는 해당 부분을 메소드로 빼는 것입니다. getStartDateAndEndDate() 메소드를 만들고, 반환시에는 startDate, endDate를 필드로 갖는 클래스를 반환하여, 혹은 Pair를 반환하여 startDate, endDate를 구하는 것입니다. 이러면 선언과 동시에 할당이 일어날 것이에요
두번째는 java 17부터는 switch문에서 반환이 가능합니다. 반환값을 이용해서 startDate, endDate를 구하면 선언과 할당을 동시에 할수 있을거에요 : https://all-i-want.tistory.com/190
보니까 java 17을 쓰시더라구요, 한번 적용해보시면 좋을 것 같아요. 그리고 java 17부터는 switch문이 좀 달라졌습니다. 변경된 switch문을 적용해보셔도 좋을것 같네요

jooyeon0907 replied:
오 switch 문을 바로 반환할 수 도 있네요! 말씀해주신 방식으로 수정하여 올리겠습니다 감사합니다!

mingming-mentor commented on src/main/java/com/zerobase/timate/service/ScheduleService.java:
```diff
@@ -0,0 +1,128 @@
+
+		List<Schedule> schedules = scheduleRepository.findByCalendarIdAndStartDateBetween(
+			calendarId, startDate.atStartOfDay(), endDate.atTime(23, 59, 59));
+		if (period.equals("daily")) {
```
여기서도 위의 switch문에서처럼 toLowerCase가 붙어야할것 같네요. 그리고 가능하면 NPE를 피하기 위해 일반적으로
"daily".equals(period) 로 많이 사용합니다. const 값을 기준으로 메소드를 실행시키는 거죠.

mingming-mentor reviewed (approved):
수고 많으셨습니다

jooyeon0907 committed:
fix: 피드백 반영하여 코드 수정
- @JsonProperty 제거
- 기간별 일정 조회 수정:
  - period 필드를 String에서 enum 타입으로 변경하여 비교
  - startDate, endDate 계산 로직을 메소드로 분리하여 Pair로 반환
  - java 17의 switch 문을 사용하여 코드 간결화
- Pair 사용을 위해 commons-lang3 라이브러리 추가

jooyeon0907 committed:
Merge remote-tracking branch 'origin/feature/schedule' into feature/schedule

# Conflicts:
#	src/main/java/com/zerobase/timate/controller/ScheduleController.java
#	src/main/java/com/zerobase/timate/dto/ScheduleDto.java
#	src/main/java/com/zerobase/timate/service/ScheduleService.java
